utils/patientmatching/rule_based_patient_matcher.bal

```diff
@@ -0,0 +1,261 @@
+        }
+
+        if <int>score == rulesRecord.fhirpathArray.length() {
+            response.setJsonPayload(true);
+        }
+
+        return response;
+
+    }
+
+    public isolated function matchPatients(r4:Patient sourcePatient, json config) returns error|http:Response {
+        record {}|error|() patient = self.getMatchingPatients(sourcePatient, config);
+        if patient is () {
+            http:Response response = new;
+            response.setJsonPayload("No matching patient found");
+            return response;
+        }
+        if patient is error {
+            return createPatientMatchingError("Error in getting patients from the master patient index database");
+        } else {
+            http:Response response = new;
+            map<anydata> mapResult = <map<anydata>>patient.get("value");
+            response.setJsonPayload(mapResult.toJson());
+            return response;
+
+        }
+    }
+
+    public isolated function getMpiDbClient(json config) returns sql:Client|error {
+
+        MPIDbConfig|error dbConfig = self.getMPIConfigData(config);
+        if dbConfig is error {
+            return createPatientMatchingError("Error in getting database configuration details from patientMatcherConfig.json file");
+        }
+        mysql:Client dbClient = check new (dbConfig.host, dbConfig.username, dbConfig.password, dbConfig.database, dbConfig.port);
+        return dbClient;
+
+    }
+
+    isolated function getMatchingPatients(r4:Patient patient, json config) returns record {}|error|() {
+        stream<record {}, sql:Error?>|error dbPatientStream = self.getMPIData(patient, check self.getPatientMatcherRuleData(config), config);
+        if dbPatientStream is error {
+            return createPatientMatchingError("Error in getting patient details stream from the master patient index database");
+
+        }
+        while (true) {
+            record {}|()|sql:Error? nextRawInDb = dbPatientStream.next();
+            if nextRawInDb is () {
+                return ();
+            }
+            if nextRawInDb is record {} {
+                return nextRawInDb;
+            } else {
+                return createPatientMatchingError(nextRawInDb.message());
+            }
+        }
+    }
+
+    isolated function getMPIData(r4:Patient patient, RulesRecord rulesTable, json config) returns stream<record {}, sql:Error?>|error {
+
+        sql:Client|error dbClient = self.getMpiDbClient(config);
+
+        if dbClient is error {
+            return createPatientMatchingError("Error in getting database client");
+        }
+        error|string qry = self.getSQLQuery(patient, rulesTable, config);
+        if qry is error {
+            return createPatientMatchingError("Error in getting SQL query from patientMatcherConfig.json file");
+        }
+        sql:ParameterizedQuery queryString = ``;
+        queryString.strings = [qry];
+        stream<record {}, sql:Error?> resultStream = dbClient->query(queryString);
```

Comment:
Can we use more specific type here? eg: https://ballerina.io/learn/by-example/mysql-query-operation/

Reply:
Yeah. As we are unable to know the structure of the Patient Record coming from Master Patient Index. I leave that as `record{}`

Reply:
Okay use `map<anydata>` then insted of `record {}` since both are the same

Reply:
Not resolved?

Reply:
Cannot use `map<anydata>` due to https://github.com/ballerina-platform/ballerina-lang/issues/41008